src/conversions/osstr.rs

```diff
@@ -0,0 +1,230 @@
+use crate::types::PyString;
+#[cfg(windows)]
+use crate::PyErr;
+use crate::PyNativeType;
+use crate::{
+    ffi, AsPyPointer, FromPyObject, IntoPy, PyAny, PyObject, PyResult, PyTryFrom, Python,
+    ToPyObject,
+};
+use std::borrow::Cow;
+use std::ffi::{OsStr, OsString};
+#[cfg(not(windows))]
+use std::os::raw::c_char;
+
+// TODO: move to other module to prevent accidentally circumventing the new function?
+#[cfg(windows)]
+struct DropGuard<T>(*mut T);
```

Comment:
I don't think it's good to abstract such a specific operation.
In this struct, `*mut T` *implicitly* must be a pointer in the Python heap without any reference to Python objects, so it's unsafe to construct this object.
So I recommend making this a more limited one without `new` in the function scope if you like this solution much more than `Vec.`

Reply:
Agreed, this type is super `unsafe` so I'd rather it was inside the function too. That way only the function can misuse it.